bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/ThreadFactoryBuilder.java

```diff
@@ -0,0 +1,181 @@
+     * @return this {@link ThreadFactoryBuilder} instance
+     */
+    public ThreadFactoryBuilder withUncaughtExceptionHandler(
+            @Nullable UncaughtExceptionHandler uncaughtExceptionHandler) {
+        this.uncaughtExceptionHandler = uncaughtExceptionHandler;
+        return this;
+    }
+
+    /**
+     * Sets the priority to be set for created threads. Must be a valid thread priority, as indicated by
+     * {@link Thread#MIN_PRIORITY} and {@link Thread#MAX_PRIORITY}. If set to null, the built {@link ThreadFactory} will
+     * not set a priority. Defaults to null.
+     *
+     * @param priority The priority to be used for created threads.
+     * @return this {@link ThreadFactoryBuilder} instance
+     */
+    public ThreadFactoryBuilder withPriority(@Nullable Integer priority) {
+        if (priority != null && priority.intValue() < Thread.MIN_PRIORITY) {
+            throw new IllegalArgumentException(String.format(
+                    "The provided priority %d is below the minimal thread priority %d", priority, Thread.MIN_PRIORITY));
+        }
+
+        if (priority != null && priority.intValue() > Thread.MAX_PRIORITY) {
+            throw new IllegalArgumentException(String.format(
+                    "The provided priority %d is above the maximal thread priority %d", priority, Thread.MAX_PRIORITY));
+        }
+
+        this.priority = priority;
+        return this;
+    }
+
+    /**
+     * Builds the {@link ThreadFactory}, configuring it as specified during the build.
+     *
+     * @return the {@link ThreadFactory}
+     */
+    public ThreadFactory build() {
+        ThreadFactory wrappedThreadFactory = this.wrappedThreadFactory;
+        if (wrappedThreadFactory == null) {
+            wrappedThreadFactory = defaultThreadFactory();
+        }
+
+        return ThreadFactoryBuilder.build(wrappedThreadFactory, namePrefix, daemonThreads, uncaughtExceptionHandler,
+                priority);
+    }
+
+    private static ThreadFactory build(ThreadFactory wrappedThreadFactory, @Nullable String namePrefix,
+            boolean daemonThreads, @Nullable UncaughtExceptionHandler uncaughtExceptionHandler,
+            @Nullable Integer priority) {
+
+        return new ThreadFactory() {
+            AtomicInteger threadCounter = new AtomicInteger(1);
+
+            @Override
+            public Thread newThread(@Nullable Runnable runnable) {
+                Thread thread = wrappedThreadFactory.newThread(runnable);
+
+                if (namePrefix != null) {
+                    thread.setName(String.format("ESH-%s-%d", namePrefix, threadCounter.getAndIncrement()));
+                }
+
+                thread.setDaemon(daemonThreads);
+
+                if (uncaughtExceptionHandler != null) {
+                    thread.setUncaughtExceptionHandler(uncaughtExceptionHandler);
+                }
+
+                if (priority != null) {
+                    thread.setPriority(priority.intValue());
```

Comment:
Shouldn't the unboxing done automatically here (https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html)?

Reply:
Yes, the compiler should automatically unbox. I like being more explicit here, but will change to automatic unboxing :)